src/libclipper/src/selection_policies.cpp

```diff
@@ -1,569 +1,117 @@
-       it != state.model_map_.end() && rand_num >= 0; ++it) {
-    rand_num -= it->second["weight"] / state.weight_sum_;
-    selected_model = it->first;
-  }
-  return selected_model;
-}
-
-std::vector<PredictTask> Exp3Policy::select_predict_tasks(
-    BanditPolicyState state, Query query, long query_id) {
-  auto selected_model = select(state);
-  auto task = PredictTask(query.input_, selected_model, 1.0, query_id,
-                          query.latency_micros_);
-  std::vector<PredictTask> tasks{task};
-  return tasks;
-}
-
-Output Exp3Policy::combine_predictions(BanditPolicyState /*state*/,
-                                       Query /*query*/,
-                                       std::vector<Output> predictions) {
-  if (predictions.empty()) {
-    std::cout << "No predictions to combine" << std::endl;
-    return Output(0.0, {});
-  }
-  return predictions.front();
-}
+DefaultOutputSelectionState::DefaultOutputSelectionState(
+    std::string serialized_state)
+    : default_output_(deserialize(serialized_state)) {}
 
-std::pair<std::vector<PredictTask>, std::vector<FeedbackTask>>
-Exp3Policy::select_feedback_tasks(BanditPolicyState& state,
-                                  FeedbackQuery feedback, long query_id) {
-  // Predict Task
-  auto selected_model = select(state);
-  auto predict_task =
-      PredictTask(feedback.feedback_.input_, selected_model, -1, query_id, -1);
-  std::vector<PredictTask> predict_tasks{predict_task};
-  // Feedback Task
-  std::vector<FeedbackTask> feedback_tasks;
-
-  return make_pair(predict_tasks, feedback_tasks);
+std::string DefaultOutputSelectionState::serialize() const {
+  rapidjson::Document d;
+  d.SetObject();
+  json::add_double(d, "y_hat", default_output_.y_hat_);
+  return json::to_json_string(d);
 }
-
-BanditPolicyState Exp3Policy::process_feedback(
-    BanditPolicyState state, Feedback feedback,
-    std::vector<Output> predictions) {
-  if (predictions.empty()) {  // Edge case
-    std::cout << "No predictions, so can't update state." << std::endl;
-    return state;
-  }
-
-  // Compute loss and find which model to update
-  auto loss = std::abs(predictions.front().y_hat_ - feedback.y_);
-  auto model_id = predictions.front().models_used_.front();
-  // Update arm weight and weight_sum
-  auto s_i = state.model_map_[model_id]["weight"];
-  if (s_i != 0) {
-    auto update = exp(-eta * loss / (s_i / state.weight_sum_));
-    state.model_map_[model_id]["weight"] = s_i * update;
-    state.set_weight_sum(state.weight_sum_ - s_i +
-                         state.model_map_[model_id]["weight"]);
-  }
-  return state;
-}
-
-std::string Exp3Policy::serialize_state(BanditPolicyState state) {
-  return state.serialize();
+std::string DefaultOutputSelectionState::get_debug_string() const {
+  rapidjson::Document d;
+  d.SetObject();
+  json::add_double(d, "y_hat", default_output_.y_hat_);
+  std::vector<std::string> empty_vec;
+  json::add_string_array(d, "models_used", empty_vec);
```

Comment:
Not really. The debug string here is what the default output would be. An output is the pair <y_hat, models_used> and if the default output is being returned, no models were used. Debug string format is purposely left up to the user.

Also, next up on the TODO list is [CLIPPER-107](https://clipper.atlassian.net/browse/CLIPPER-107), which will change Output to be a json string. So this output will be refactored anyway.